### PR TITLE
Support for addon parameters before installation

### PR DESCRIPTION
**Pull request by oharan2**

### CODE REVIEW

File: ocm_python_wrapper/cluster.py [hunk 1]
```diff
@@ -304,8 +303,8 @@ def _get_required_cluster_parameters(_addon_parameters):
             for param in _addon_parameters["items"]:
                 param_conditions = [
                     con["data"]
-                    for con in param["conditions"]
-                    if param["required"] is True and con["resource"] == "cluster"
+                    for con in (param.get("conditions") or [])
+                    if param.get("required") and con["resource"] == "cluster"
```
Comment (oharan2) [inline ocm_python_wrapper/cluster.py:307]: If `conditions` is not a key nor `required` is True - `param_conditions = []`

File: ocm_python_wrapper/cluster.py [hunk 2]
```diff
@@ -262,26 +263,112 @@ def addon_info(self):
...
+                example:
+                _required_parameters = {'cidr-range': {'default_value': '10.1.0.0/26'}, 'addon_parameter': {
+                        'default_value': ''}, ..}
+
+            """
+            _required_parameters = {}
+
+            for param in _addon_parameters["items"]:
+                param_conditions = [
+                    con["data"]
```
Comment (rnetser) [inline ocm_python_wrapper/cluster.py:305]:
```suggestion
                    condition["data"]
```

File: ocm_python_wrapper/cluster.py [hunk 3]
```diff
@@ -262,26 +263,112 @@ def addon_info(self):
...
+                            _condition_value=condition_value,
+                        ):
+                            break
+                    else:
+                        _required_parameters[param["id"]] = {
+                            "default_value": param.get("default_value")
+                        }
+            return _required_parameters
+
+        def _check_param_conditions(_clusters_dict, _condition, _condition_value):
```
Comment (rnetser) [inline ocm_python_wrapper/cluster.py:323]: move before the function's code; this can be a static method

File: ocm_python_wrapper/cluster.py [hunk 4]
```diff
@@ -262,26 +263,112 @@ def addon_info(self):
...
+            """
+            Check if parameter conditions met with cluster configuration
+
+            Args:
+                _clusters_dict (dict): Cluster instance dict
+                _condition (str): Condition key to check
+                _condition_value (str): Condition expected value
+
+            Returns:
+                Bool: True if cluster instance match with condition, else false
```
Comment (rnetser) [inline ocm_python_wrapper/cluster.py:333]: `else False`

**Review by rnetser — CHANGES_REQUESTED** [review]

**Review by rnetser — DISMISSED** [review]

File: ocm_python_wrapper/cluster.py [hunk 5]
```diff
@@ -262,26 +263,91 @@ def addon_info(self):
...
         ).to_dict()
 
-    def validate_addon_parameters(self, parameters):
+    def validate_and_update_addon_parameters(
+        self, user_parameters, use_api_defaults=True
+    ):
+        """Validate and update user input parameters against API's conditions and requirements.
+
+        Args:
+            user_parameters (List): User parameters
```
Comment (myakove) [inline ocm_python_wrapper/cluster.py:272]: `List` should be `list`

File: ocm_python_wrapper/cluster.py [hunk 6]
```diff
@@ -262,26 +263,91 @@ def addon_info(self):
...
+            ValueError: When a required parameter is missing,
+                        or when parameters are passed but not needed for addon.
+        """
+
+        def _get_required_cluster_parameters(_addon_parameters):
+            """Filter cluster-related ``addon parameters``
+
+            Args:
+                _addon_parameters (dict) : Addons parameters from Clusters Management
+            ``cluster_mgmt_v1_addons_addon_id`` API.
```
Comment (myakove) [inline ocm_python_wrapper/cluster.py:292]: why surround with ``cluster_mgmt_v1_addons_addon_id``

File: ocm_python_wrapper/cluster.py [hunk 7]
```diff
@@ -262,26 +263,91 @@ def addon_info(self):
             self.addon_name
         ).to_dict()
 
-    def validate_addon_parameters(self, parameters):
+    def validate_and_update_addon_parameters(
+        self, user_parameters, use_api_defaults=True
```
Comment (myakove) [inline ocm_python_wrapper/cluster.py:267]: why `user_parameters` is mandatory? 
If I have addon that have all parameters with defaults, then the user can call this without any parameters and get the defaults.

File: ocm_python_wrapper/cluster.py [hunk 8]
```diff
@@ -262,26 +263,91 @@ def addon_info(self):
...
+        Args:
+            user_parameters (List): User parameters
+                Example:
+                    parameters = [{"id": "has-external-resources", "value": "false"},
+                                    {"id": "aws-cluster-test-param", "value": "false"},]
+            use_api_defaults (bool): If true, set required parameter (which are not part of `user_parameters`) with
+                                            default value to not fail as missing.
+
+        Returns:
+            List: Updated parameters (if default values updated) to provide for installation.
```
Comment (myakove) [inline ocm_python_wrapper/cluster.py:280]: This function can return None is some cases.
Please make sure you always return list, empty one if needed.

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocm_python_wrapper/cluster.py (hunk 7 above)
Comment (oharan2) [reply ocm_python_wrapper/cluster.py:267]: Could be set with default such as user_parameters=[], but it requires to check outside if parameters are not empty, but the results are the same (pass empty list). 
If the case you suggested happens, then `user_parameters=[]` and the values will fill with default values if possible.

An indication of `user_parameters` (also an empty list) is mandatory for:
1. Check and raise if addon doesn't accept parameters but user gave parameters with values
2. The opposite case than you suggested - if all addon parameters have defaults, but the user specified a value for a parameter; then we need `user_parameters` for the check if parameter value not specified by user (and has default) - only then use default.

Comment (myakove) [reply ocm_python_wrapper/cluster.py:267]: if `user_parameters` can be empty list then it is not mandatory

**Review by myakove — CHANGES_REQUESTED** [review]

**Review by myakove — CHANGES_REQUESTED** [review]
Check my previous comments

File: ocm_python_wrapper/cluster.py (hunk 8 above)
Comment (oharan2) [reply ocm_python_wrapper/cluster.py]: Done - we can only return `user_parameters` (list) which is set as an empty list in the base case.

File: ocm_python_wrapper/cluster.py (hunk 7 above)
Comment (oharan2) [reply ocm_python_wrapper/cluster.py]: Done

File: ocm_python_wrapper/cluster.py [hunk 9]
```diff
@@ -262,26 +263,94 @@ def addon_info(self):
...
+                        ):
+                            break
+                    else:
+                        _required_parameters[param["id"]] = {
+                            "default_value": param.get("default_value")
+                        }
+            return _required_parameters
+
+        if user_parameters is None:
+            user_parameters = []
```
Comment (myakove) [inline ocm_python_wrapper/cluster.py:324]: `user_parameters = user_parameters or []`

File: ocm_python_wrapper/cluster.py [hunk 10]
```diff
@@ -262,26 +263,94 @@ def addon_info(self):
...
         ).to_dict()
 
-    def validate_addon_parameters(self, parameters):
+    def validate_and_update_addon_parameters(
+        self, user_parameters=None, use_api_defaults=True
+    ):
+        """Validate and update user input parameters against API's conditions and requirements.
+
+        Args:
+            user_parameters (list): User parameters, default is empty list.
```
Comment (myakove) [inline ocm_python_wrapper/cluster.py:272]: `default is empty list` >> default is None

File: ocm_python_wrapper/cluster.py [hunk 11]
```diff
@@ -262,26 +263,94 @@ def addon_info(self):
...
+
+        if user_parameters is None:
+            user_parameters = []
+
         _info = self.addon_info()
-        _parameters = _info.get("parameters")
-        if not _parameters and parameters:
+        addon_parameters = _info.get("parameters")
+        if not addon_parameters and user_parameters:
             raise ValueError(f"{self.addon_name} does not take any parameters")
```
Comment (myakove) [inline ocm_python_wrapper/cluster.py:329]: Please add `user_parameters` to the error

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocm_python_wrapper/cluster.py (hunk 9 above)
Comment (oharan2) [reply ocm_python_wrapper/cluster.py]: Done - used _user_parameters

Comment (oharan2) [reply ocm_python_wrapper/cluster.py]: @myakove legitimate?

**Review by myakove — APPROVED** [review]

**Pull request merged**